browser/ui/views/location_bar/brave_news_location_view.cc

```diff
@@ -10,210 +10,154 @@
-                                           base::Unretained(this)));
-
-    auto menu_button_controller = std::make_unique<views::MenuButtonController>(
-        this,
-        base::BindRepeating(&BraveNewsButtonView::ButtonPressed,
-                            base::Unretained(this)),
-        std::make_unique<views::Button::DefaultButtonControllerDelegate>(this));
-    SetButtonController(std::move(menu_button_controller));
-
-    Update();
+BraveNewsLocationView::BraveNewsLocationView(
+    Profile* profile,
+    IconLabelBubbleView::Delegate* icon_label_bubble_delegate,
+    PageActionIconView::Delegate* page_action_icon_delegate)
+    : PageActionIconView(nullptr,
+                         0,
+                         icon_label_bubble_delegate,
+                         page_action_icon_delegate,
+                         "BraveNewsFollow") {
+  SetLabel(l10n_util::GetStringUTF16(IDS_BRAVE_NEWS_ACTION_VIEW_TOOLTIP));
+
+  last_contents_ = GetWebContents();
+  if (last_contents_) {
+    BraveNewsTabHelper::FromWebContents(last_contents_)->AddObserver(this);
   }
-  ~BraveNewsButtonView() override = default;
-  BraveNewsButtonView(const BraveNewsButtonView&) = delete;
-  BraveNewsButtonView& operator=(const BraveNewsButtonView&) = delete;
-
-  void Update() {
-    if (!should_show_.GetValue() || !news_enabled_.GetValue() ||
-        !opted_in_.GetValue()) {
-      SetVisible(false);
-      return;
-    }
 
-    auto* contents = tab_strip_->GetActiveWebContents();
-    bool subscribed = false;
-    bool has_feeds = false;
-    absl::optional<BraveNewsTabHelper::FeedDetails> feed;
+  should_show_.Init(brave_news::prefs::kShouldShowToolbarButton,
+                    profile->GetPrefs(),
+                    base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                        base::Unretained(this)));
+  opted_in_.Init(brave_news::prefs::kBraveTodayOptedIn, profile->GetPrefs(),
+                 base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                     base::Unretained(this)));
+  news_enabled_.Init(brave_news::prefs::kNewTabPageShowToday,
+                     profile->GetPrefs(),
+                     base::BindRepeating(&BraveNewsLocationView::UpdateImpl,
+                                         base::Unretained(this)));
+
+  Update();
+}
+
+BraveNewsLocationView::~BraveNewsLocationView() = default;
+
+views::BubbleDialogDelegate* BraveNewsLocationView::GetBubble() const {
+  return bubble_view_;
+}
 
+void BraveNewsLocationView::UpdateImpl() {
+  auto* contents = GetWebContents();
+  BraveNewsTabHelper* tab_helper =
+      contents ? BraveNewsTabHelper::FromWebContents(contents) : nullptr;
+
+  // When the active tab changes, subscribe to notification when
+  // it has found a feed, and update display with current state.
+  if (contents != last_contents_) {
+    // Unobserve old Tab
+    if (last_contents_) {
```

Comment:
I'm wondering if the last_contents could be dangled when web contents is destroyed. In what order does it work when we're closing the active tab? if it works likes `1. activate another contents first, 2. and then destroy web contents` then things will be fine. But I still feel like it could be too fragile. Should we observe last_contents_'s destruction and mark |last_contents_| nullptr when it's destroyed?

Reply:
At first I thought it's better to have a `WeakPtr` to `WebContents` instead of setting up an observer simply for destruction. 
But then I decided to try just holding a ScopedObservation to the TabHelper and switch or reset the observation. I think both the TabHelper's `ObserverList` combined with a `ScopedObservation` allows for either the observer or the observed to be destroyed first with no risk of crash should lifetimes change or get notified in a different order. What do you think? 42df97a03fcedbb2d26d849e9d7d71c39c07537b

Reply:
Actually I see a crash in some browser tests, investigating...

Reply:
Still needed to remove the observer when the WebContents is getting destroyed 8f9add61cf2a0e8682570d402c0d3f45fd572022

Reply:
Still needed to remove the observer when the WebContents is getting destroyed 8f9add61cf2a0e8682570d402c0d3f45fd572022

Reply:
Still needed to remove the observer when the WebContents is getting destroyed 8f9add61cf2a0e8682570d402c0d3f45fd572022